plugins/action_send_notification/src/sendNotification.ts

```diff
@@ -0,0 +1,127 @@
+    sendEmail = true,
+  } = parameters
+
+  const {
+    environmentData: { appRootFolder, filesFolder },
+  } = applicationData as ActionApplicationData
+
+  const transporter = nodemailer.createTransport({
+    host,
+    port,
+    secure,
+    auth: {
+      user,
+      pass: process.env.SMTP_PASSWORD,
+    },
+  })
+
+  try {
+    const emailAddressString = stringifyEmailRecipientsList(email)
+
+    // Create notification database record
+    console.log('Creating notification...')
+    let notificationResult
+    notificationResult = await db.createNotification({
+      userId,
+      applicationId: applicationData?.applicationId,
+      emailAddressString,
+      subject,
+      message,
+      attachments,
+    })
+
+    // Send email
+    if (sendEmail) {
+      console.log('Sending email...')
+      // Note: Using "any" type as imported @types defintions is incorrect, doesn't recognise some fields on "SentMessageInfo" type
+      const emailResult: any = await transporter.sendMail({
+        from: `${fromName} <${fromEmail}>`,
+        to: emailAddressString,
+        subject,
+        text: message,
+        html: marked(message),
+        attachments: await prepareAttachments(attachments, appRootFolder, filesFolder),
+      })
+
+      // Update notification table with email sent confirmation
+      if (emailResult?.response.match(/250 OK.*/)) {
+        console.log(`Email successfully sent to: ${emailResult.envelope.to}\n`)
+        notificationResult = await db.notificationEmailSent(notificationResult.id)
+      }
+    }
+
+    return {
+      status: ActionQueueStatus.Success,
+      error_log: '',
+      output: { notification: notificationResult },
+    }
+  } catch (error) {
+    console.log(error.message)
+    return {
+      status: ActionQueueStatus.Fail,
+      error_log: error.message,
+    }
+  }
+}
+
+export default sendNotification
+
+const stringifyEmailRecipientsList = (emailAddresses: string | string[]): string => {
+  if (!Array.isArray(emailAddresses)) return emailAddresses
+  return emailAddresses.join(', ')
+}
+
+/*
+Takes an array of file paths in multiple formats and builds them into 
+shape expected by sendMail method.
+Input elements must be one of:
+- <string> uniqueId of file (we will look up the path and filename)
+- already formatted sendMail object, with the following properties:
+{
+  path: <full path/url of file>
+  filename: <name to give attached file>
+}
+- TO-DO: Handle more types of input format (e.g. raw path/url strings)
+*/
+const prepareAttachments = async (
+  attachments: string[] | Attachment[],
+  appRootFolder: string,
+  filesFolder: string
+): Promise<Attachment[]> => {
+  const attachmentObjects = []
+  for (const file of attachments) {
+    if (typeof file === 'object') {
+      if (!file?.path || !file?.filename) throw new Error('Invalid attachment')
+      attachmentObjects.push(file)
+    } else {
+      const { original_filename, file_path } = await getFilePath(file)
```

Comment:
Is this `original_filename` that you said would be good to get from uploaded files that aren't getting the proper original name from file but a random number instead?

Reply:
It's okay in this context, as the emailer can set the filename of the outgoing file, which is what's happening here. The problem I was referring to is specifically to do with the "sendFile" method used by the `/file` download endpoint.

Reply:
So when we upload files?

Reply:
Sorry, I'm not following. Hit me up on Telegram if you want further clarification.